File: backend/src/modules/versions/versionService.js
import { logAction } from '../audit/auditService.js';
import { v4 as uuidv4 } from 'uuid';
import versionModel from './versionModel.js';
import awsStorageService from '../../services/awsStorageService.js';

class VersionService {
    async createNewVersion(cibleId, type, content, versionMetadata, utilisateur_id) {
        try {
            const versionId = uuidv4();

            // Convertir le contenu en Buffer si nécessaire
            let contentBuffer;
            if (Buffer.isBuffer(content)) {
                contentBuffer = content;
            } else if (typeof content === 'string') {
                contentBuffer = Buffer.from(content);
            } else if (typeof content === 'object') {
                contentBuffer = Buffer.from(JSON.stringify(content));
            } else {
                throw new Error('Type de contenu non supporté');
            }

            // Upload vers S3
            const s3Result = await awsStorageService.uploadVersion(contentBuffer, versionId, cibleId, type);

            // Insérer la nouvelle version dans la base de données
            const newVersion = await versionModel.createVersion({
                id: versionId,
                cible_id: cibleId,
                type: type,
                version_number: versionMetadata.version_number,
                storage_path: s3Result.key, // Clé S3 au lieu du chemin local
                metadata: {
                    ...versionMetadata,
                    s3Location: s3Result.location,
                    s3Size: s3Result.size
                },
                created_by: utilisateur_id
            });

            // Journaliser l'action
            if (utilisateur_id) {
                await logAction(
                    utilisateur_id,
                    'create_version',
                    'version',
                    versionId,
                    {
                        version_id: versionId,
                        type: type,
                        s3_location: s3Result.location,
                        date: new Date()
                    }
                );
            }

            return newVersion;

        } catch (error) {
            console.error('Erreur lors de la création de la version:', error);
            throw error;
        }
    }

    async getVersionHistory(cibleId, type) {
        try {
            console.log('Récupération de l\'historique des versions pour cible_id:', cibleId, 'et type:', type);
            const versions = await versionModel.getVersionsByCibleId(cibleId, type);
            console.log('Versions trouvées:', versions);
            return versions;
        } catch (error) {
            console.error('Erreur lors de la récupération de l\'historique:', error);
            throw error;
        }
    }

    async getVersionContent(versionId) {
        const version = await versionModel.getVersionById(versionId);
        if (!version) {
            return null;
        }

        try {
            // Télécharger le contenu depuis S3
            const content = await awsStorageService.downloadVersion(version.storage_path);
            
            // Tenter de parser le contenu comme JSON si c'est un objet
            try {
                const jsonContent = JSON.parse(content.toString());
                return {
                    ...version,
                    content: jsonContent
                };
            } catch {
                // Si ce n'est pas du JSON, retourner le contenu brut
                return {
                    ...version,
                    content: content
                };
            }
        } catch (error) {
            console.error('Erreur lors de la lecture du contenu de la version:', error);
            throw error;
        }
    }

    async deleteVersion(versionId) {
        const version = await versionModel.getVersionById(versionId);
        if (!version) {
            return null;
        }

        try {
            // Supprimer le fichier de S3
            await awsStorageService.deleteVersion(version.storage_path);

            // Supprimer l'enregistrement de la base de données
            return await versionModel.deleteVersion(versionId);
        } catch (error) {
            console.error('Erreur lors de la suppression de la version:', error);
            throw error;
        }
    }

    async compareVersions(versionId1, versionId2) {
        const [version1, version2] = await Promise.all([
            this.getVersionContent(versionId1),
            this.getVersionContent(versionId2)
        ]);

        if (!version1 || !version2) {
            throw new Error('Une ou plusieurs versions non trouvées');
        }

        // Comparer les métadonnées
        const metadataDiff = this.compareMetadata(version1.metadata, version2.metadata);

        // Comparer le contenu
        const contentDiff = version1.content.equals(version2.content) ? null : 'Contenu différent';

        return {
            metadataDiff,
            contentDiff
        };
    }

    compareMetadata(metadata1, metadata2) {
        const diff = {};
        const allKeys = new Set([...Object.keys(metadata1), ...Object.keys(metadata2)]);

        for (const key of allKeys) {
            if (JSON.stringify(metadata1[key]) !== JSON.stringify(metadata2[key])) {
                diff[key] = {
                    old: metadata1[key],
                    new: metadata2[key]
                };
            }
        }

        return Object.keys(diff).length > 0 ? diff : null;
    }

    // Générer une URL signée pour télécharger une version
    async getVersionDownloadUrl(versionId, expiresIn = 3600) {
        const version = await versionModel.getVersionById(versionId);
        if (!version) {
            return null;
        }

        try {
            return await awsStorageService.getSignedDownloadUrl(version.storage_path, expiresIn);
        } catch (error) {
            console.error('Erreur lors de la génération de l\'URL de téléchargement:', error);
            throw error;
        }
    }
}

export default new VersionService(); 

File: arkiva/lib/services/version_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:arkiva/config/api_config.dart';

class VersionService {
  static String get baseUrl => ApiConfig.baseUrl;

  // Créer une nouvelle version
  static Future<Map<String, dynamic>> createVersion({
    required String token,
    required int cibleId,
    required String type,
    String? versionNumber,
    String? description,
  }) async {
    try {
      print('[API] POST $baseUrl/api/versions');
      print('[API] Headers: {Content-Type: application/json, Authorization: Bearer $token}');
      print('[API] Body: {cible_id: $cibleId, type: $type, version_number: $versionNumber, description: $description}');
      final response = await http.post(
        Uri.parse('$baseUrl/api/versions'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({
          'cible_id': cibleId,
          'type': type,
          if (versionNumber != null) 'version_number': versionNumber,
          if (description != null) 'description': description,
        }),
      );

      if (response.statusCode == 201) {
        return json.decode(response.body);
      } else {
        final errorBody = json.decode(response.body);
        throw Exception(errorBody['error'] ?? 'Erreur lors de la création de la version');
      }
    } catch (e) {
      throw Exception('Erreur de connexion: $e');
    }
  }

  // Obtenir l'historique des versions
  static Future<List<Map<String, dynamic>>> getVersionHistory({
    required String token,
    required int cibleId,
    required String type,
  }) async {
    try {
      print('[API] GET $baseUrl/api/versions/cible/history?cible_id=$cibleId&type=$type');
      print('[API] Headers: {Authorization: Bearer $token}');
      final response = await http.get(
        Uri.parse('$baseUrl/api/versions/cible/history?cible_id=$cibleId&type=$type'),
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return List<Map<String, dynamic>>.from(data['versions'] ?? []);
      } else {
        final errorBody = json.decode(response.body);
        throw Exception(errorBody['error'] ?? 'Erreur lors de la récupération de l\'historique');
      }
    } catch (e) {
      throw Exception('Erreur de connexion: $e');
    }
  }

  // Obtenir le contenu d'une version
  static Future<Map<String, dynamic>> getVersionContent({
    required String token,
    required String versionId,
  }) async {
    try {
      print('[API] GET $baseUrl/api/versions/$versionId/content');
      print('[API] Headers: {Authorization: Bearer $token}');
      final response = await http.get(
        Uri.parse('$baseUrl/api/versions/$versionId/content'),
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        final errorBody = json.decode(response.body);
        throw Exception(errorBody['error'] ?? 'Erreur lors de la récupération du contenu');
      }
    } catch (e) {
      throw Exception('Erreur de connexion: $e');
    }
  }

  // Obtenir l'URL de téléchargement d'une version
  static Future<String> getVersionDownloadUrl({
    required String token,
    required String versionId,
    int expiresIn = 3600,
  }) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/versions/$versionId/download-url?expiresIn=$expiresIn'),
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['download_url'];
      } else {
        final errorBody = json.decode(response.body);
        throw Exception(errorBody['error'] ?? 'Erreur lors de la récupération de l\'URL de téléchargement');
      }
    } catch (e) {
      throw Exception('Erreur de connexion: $e');
    }
  }

  // Comparer deux versions
  static Future<Map<String, dynamic>> compareVersions({
    required String token,
    required String versionId1,
    required String versionId2,
  }) async {
    try {
      print('[API] POST $baseUrl/api/versions/compare');
      print('[API] Headers: {Content-Type: application/json, Authorization: Bearer $token}');
      print('[API] Body: {version_id_1: $versionId1, version_id_2: $versionId2}');
      final response = await http.post(
        Uri.parse('$baseUrl/api/versions/compare'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: json.encode({
          'version_id_1': versionId1,
          'version_id_2': versionId2,
        }),
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        final errorBody = json.decode(response.body);
        throw Exception(errorBody['error'] ?? 'Erreur lors de la comparaison des versions');
      }
    } catch (e) {
      throw Exception('Erreur de connexion: $e');
    }
  }

  // Supprimer une version
  static Future<void> deleteVersion({
    required String token,
    required String versionId,
  }) async {
    try {
      print('[API] DELETE $baseUrl/api/versions/$versionId');
      print('[API] Headers: {Authorization: Bearer $token}');
      final response = await http.delete(
        Uri.parse('$baseUrl/api/versions/$versionId'),
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode != 200 && response.statusCode != 204) {
        final errorBody = json.decode(response.body);
        throw Exception(errorBody['error'] ?? 'Erreur lors de la suppression de la version');
      }
    } catch (e) {
      throw Exception('Erreur de connexion: $e');
    }
  }
} 